### DISCUSSION TITLE
Advanced Types - EntityCreateArgs Type with Includes

### DISCUSSION THREAD

**Discussion by ctsstc** [prisma/prisma#21037] [Q&A]
### Question

I see that there's some great documentation on Advanced Types here: https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types#problem-using-variations-of-the-generated-model-type

I have not been able to figure out though how to utilize the `Prisma.EntityNameEntityCreateArgs` to also explicitly include `include`s for my typing. Do I need to create a manual union type, is there something I could pass into the generic for it? I've not been able to figure this out, but i imagine this must be possible.

## Example

I have a converter function that goes from an internally defined interface that converts to the `Prisma.EntityNameEntityCreateArgs` but if the typing does not relay that the `include` is provided on the args it's seen as optional when using that type and thus the query performed does not return the expected return type via typing (in reality it would work just fine, but it will not pass the TSC).

```ts
export class SomeConverter {
  public static someCreateParamsToEntity(createParams: SomeCreateParams): Prisma.SomeEntityCreateArgs {
    return {
      data: {
        userId: createParams.userId,
      },
      include: {
        category: true,
        merchant: true,
      },
    }
  }
}

const createEnittyArgs = SomeConverter.someCreateParamsToEnity(myCreateParams)
// TS will not show that this has the expected category and merchant information being returned since it's "optional".
const newEnitity = prisma.someEntity.create(createEnittyArgs)
```

## Temporary Get-Around ?

I'm hoping there's some helper available or a better way than this?

```ts
const categoryAndMerchantInclude = {
  category: true,
  merchant: true,
}
type CategoryAndMerchantInclude = typeof categoryAndMerchantInclude

export class SomeConverter {
  public static someCreateParamsToEntity(
    createParams: SomeCreateParams,
  ): Prisma.SomeEntityCreateArgs & { include: CategoryAndMerchantInclude } {
    return {
      data: {
        userId: createParams.userId,
      },
      include: categoryAndMerchantInclude,
    }
  }
}

const createEnittyArgs = SomeConverter.someCreateParamsToEnity(myCreateParams)
// Now TS will know that category and merchant is available on the newly created entitty.
const newEnitity = prisma.someEntity.create(createEnittyArgs)
```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by millsp** [comment]
Hey @ctsstc, have you tried using the validator?
https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/prisma-validator

**Reply by ctsstc** [reply]
I utilized the Prisma Validator for the return lookup type, but I couldn't figure out how to do it for the creation so that it fulfils the return type.

**Reply by millsp** [reply]
Could you please share a simplified example of what you tried?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
